core/server/master/src/main/java/alluxio/master/file/meta/SimpleInodeLockList.java

```diff
@@ -0,0 +1,381 @@
+    Preconditions.checkState(!endsInInode());
+    Preconditions.checkState(!mEntries.isEmpty());
+    Preconditions.checkState(endsInWriteLock());
+
+    if (endsInMultipleWriteLocks()) {
+      lockInode(inode, LockMode.WRITE);
+      return;
+    }
+
+    EdgeEntry edgeEntry = createEdgeEntry(lastEdge(), LockMode.READ);
+    InodeEntry inodeEntry = createInodeEntry(inode, mode);
+    removeLastEntry();
+    addEntry(edgeEntry);
+    addEntry(inodeEntry);
+  }
+
+  /**
+   * Locks an edge and creates and entry without adding the entry to the lock list yet.
+   *
+   * @return the edge entry
+   */
+  private EdgeEntry createEdgeEntry(Edge edge, LockMode mode) {
+    LockResource lock = mInodeLockManager.lockEdge(edge, mode);
+    return new EdgeEntry(lock, edge, mode);
+  }
+
+  /**
+   * Locks an inode and creates and entry without adding the entry to the lock list yet.
+   *
+   * @return the inode entry
+   */
+  private InodeEntry createInodeEntry(Inode inode, LockMode mode) {
+    LockResource lock = mInodeLockManager.lockInode(inode, mode);
+    return new InodeEntry(lock, inode, mode);
+  }
+
+  /**
+   * If the lock list is write locked, returns LockMode.WRITE. Otherwise, returns the given mode.
+   *
+   * This helps us preserve the invariant that there is never a READ lock following a WRITE lock.
+   *
+   * @param mode a lock mode
+   * @return the mode
+   */
+  private LockMode nextLockMode(LockMode mode) {
+    if (endsInWriteLock()) {
+      return LockMode.WRITE;
+    }
+    return mode;
+  }
+
+  /**
+   * Adds an entry to mEntries.
+   *
+   * @param entry the entry to add
+   */
+  private void addEntry(Entry entry) {
+    if (!endsInWriteLock() && entry.getMode() == LockMode.READ) {
+      mFirstWriteLockIndex++;
+    }
+    mEntries.add(entry);
+  }
+
+  /**
+   * Removes and unlocks the last entry from mEntries.
+   */
+  private void removeLastEntry() {
+    mEntries.remove(mEntries.size() - 1).getLock().close();
+    if (mFirstWriteLockIndex > mEntries.size()) {
+      mFirstWriteLockIndex = mEntries.size();
+    }
+  }
+
+  @Override
+  public LockMode getLockMode() {
+    return endsInWriteLock() ? LockMode.WRITE : LockMode.READ;
+  }
+
+  @Override
+  public List<Inode> getLockedInodes() {
+    return mEntries.stream()
+        .filter(e -> e instanceof InodeEntry)
+        .map(e -> ((InodeEntry) e).getInode())
+        .collect(Collectors.toList());
+  }
+
+  @Override
+  public Inode get(int index) {
+    int entryIndex = mEntries.get(0) instanceof InodeEntry ? index * 2 : index * 2 + 1;
+    return ((InodeEntry) mEntries.get(entryIndex)).getInode();
+  }
+
+  @Override
+  public int numInodes() {
+    if (mEntries.isEmpty()) {
+      return 0;
+    }
+    if (mEntries.get(0) instanceof InodeEntry) {
+      return (mEntries.size() + 1) / 2;
+    }
+    return mEntries.size() / 2;
+  }
+
+  @Override
+  public boolean isEmpty() {
+    return mEntries.isEmpty();
```

Comment:
can `mEntries` have just a single edge (and no inode), and if so, is that considered empty?

Reply:
that's a valid state. I would not consider it empty since the list contains a locked edge.